### ISSUE TITLE
Needed testers for new system

### ISSUE THREAD

**Issue by gusmanb** [gusmanb/PSVRFramework#42]
I've finished the new system with the service apart from the toolbox and the reduced interface, so I need people to test it.

This is a test build, is in debug mode so it will output any exception. Also, please, DON'T ASK ABOUT HEAD TRACKING, it works but there's nothing to consume it, so no, this doesn't works with steam nor the VR player.

The toolbox can work in two modes, standalone or service dependant. 
Standalone mode does that, work standalone without the need of the service, this mode is planned for people which wants only to use the cinematic mode and maybe the vr player (still no input broadcast, but may be I will add it on a future).
Service dependant uses the new (preferred) mode where it connects to the service.

The service is a Windows service, it will boot with your computer automatically so the PS VR headset will be always ready to be used by any program implementing the client. The client code is included on the PSVRFramework library, if any one wants to create a program which controls the device just needs to link to the library, create a new instance of the client and call the exposed functions like HeadsetOn or EnterCinematicMode. Also, as I'm programming the OpenVR driver I got the client on C++, I will post it when I get the driver working.

The service is self-managed, you can install it from the command line, just call "PSVRService /I" and it will be installed, /U uninstalls it, and also you can control it's state with /S /P and /R. Run "PSVRService /H" for more info.

The service should be compatible with Linux, instead of starting with "mono (executable)" it must be started with "mono-service PSVRService.exe". It will return immediately as it will start a daemon, it's execution can be controlled with kill to send signals, more info [here](https://linux.die.net/man/1/mono-service).

So, if anyone wants to test the new packages just uncompress them each one on a folder. I would need these tests:

1-Install the service, it will boot automatically (on Linux run it with mono-service).
2-Start the toolbox and try it, report any problem.
3-Go to settings on the toolbox and enable standalone mode.
4-Exit from the toolbox.
5-Uninstall service (on Linux use kill (service-proc-id)).
6-Boot up the toolbox, it will now control the device without the service, try it and report any problem.

[PSVRService.zip](https://github.com/gusmanb/PSVRFramework/files/614687/PSVRService.zip)
[PSVRToolbox.zip](https://github.com/gusmanb/PSVRFramework/files/614688/PSVRToolbox.zip)

**Comment by sidonsoft** [comment]
PSVRService is working as intended
PSVRToolbox is working as intended in standalone

sidenote: test_player_2 no longer switches to VR mode with this version of toolbox on opening video file, Headset displays loading circle but stays in Cinematic mode when video starts.

**Comment by MaVe-64** [comment]
For me there where problems (Windows 7 Home Premium 64b).
The service installs but doesn't start. If I start it manually (services.msc) I get 'Error 5 Access denied'.
Switching the user to 'Local System Account' instead of 'Network Service' account then the service starts.

Toolbox works as expected.

**Comment by gusmanb** [comment]
@MaVe-64 did you installed the service as administrator? I forgot to explain it, you need to execute it from a console as admin.

**Comment by MaVe-64** [comment]
Yes, I'm Local admin on my PC.

**Comment by CalmBSK** [comment]
Hi, my results -

Windows 10 64bit Laptop
PSVRService installed with admin rights - service fails to start - 'Error 5 Access denied'
As per MaVe-64, switching to Local instead of Network allows the service to start.
PSVRToolbox with PSVRService runs and locates PSVR.
Uninstalled PSVRService, ran PSVRToolbox and PSVR found.

Windows 7 64bit Desktop - AMD GPU
PSVRService installed with admin rights - service successfully started manually
PSVRToolbox with PSVRService runs and locates PSVR
Uninstalled PSVRService, ran PSVRToolbox and PSVR found.

For what it's worth - the Enable VR / Enable Cinematic / Recenter / Recalibrate buttons appear to do nothing for me with my setup. I believe I am in VR mode at all times and cannot enable Cinematic mode. 
The Cinematic Mode controls do not change anything for me when changed. This is with and without PSVRService installed. I appreciate this is a test build so have no idea what is and is not working but wanted to mention it just in case. I'm fully aware though that I may be doing something wrong!

Thanks for all of your hard work on this.

**Comment by gusmanb** [comment]
@CalmBSK That sounds like if you were connectiong the HMD directly to the PC instead to the black box, may be this the case?

Cheers.

**Comment by gusmanb** [comment]
@MaVe-64 Is not the same to be local administrator than executing with administrator privileges, you must right-click the console direct access and then choose "Run as administrator" and then it should work.

**Comment by MaVe-64** [comment]
I executed the command line (cmd.exe) as administrator. Then psvrservice /I.

**Comment by gusmanb** [comment]
Very strange, maybe the home edition is messing with the privileges you get and you aren't really administrator, only cause to not be able to manage the service is because you don't have these privileges, just tested on my computer, run console without explicitly selecting administrator, acces error, run with privileges, all good.

**Comment by PomanoB** [comment]
Windows 10 64bit.
Everything works well, as a service, and standalone.
I propose to run PSVRServise as a normal program without the service.
Not to do this functionality in PSVRToolbox.

**Comment by gusmanb** [comment]
@PomanoB That defeats the purpose of all this rewrite, if you want the toolbox without the service use it as standalone, but it will have no support for SteamVR and VR player.

**Comment by PomanoB** [comment]
I mean, all the same, but start not as service.
There is no need to duplicate code in PSVRToolbox.

**Comment by MaVe-64** [comment]
I can manage the service, it's the network service account that doesn't seem to have the privilege.

![service](https://cloud.githubusercontent.com/assets/22538938/20650075/4cb586f2-b4ce-11e6-954b-a8f95b3fbdef.png)

**Comment by PomanoB** [comment]
Maybe, three component? 
![image](https://cloud.githubusercontent.com/assets/853705/20650443/834d9f68-b4df-11e6-80a1-15007fac32aa.png)

**Comment by CalmBSK** [comment]
@gusmanb Yes, thank you, that was my mistake.

**Comment by gusmanb** [comment]
@PomanoB That graph is exactly the new structure, you just missed to add the VR player connected to the service and you get the exact idea. But no, it will not run if it's not a service. if you don't want to use the service just boot the toolbox in standalone mode.

**Comment by gusmanb** [comment]
@MaVe-64 I will check, I used just Network Service as this is a network service and I never had any type of problem with this (but never tried with a home edition).

Can you test something? put it as Local System Account, leave the HMD on and restart your computer, does the headset blinks signaling it's being calibrated?

If yes then there's no problem at all, I can change the account to local system.

**Comment by PomanoB** [comment]
@gusmanb Ok, I understand, thank you

**Comment by MaVe-64** [comment]
Yes, after a restart as soon as the HMD automatically connects, it blinks all leds. So it works fine.

**Comment by gusmanb** [comment]
@PomanoB Don't think that was just a preference or just me being stubborn, PS VR has a big flaw with Windows 10, if you reboot your computer and the PS VR is in cinematic mode (not sure which exact combination causes this as it doesn't happens always) Windows 10 is not able to start, the thing about setting this as a service is to allow it to gracefully shutdown the box when the O.S. is going to restart.

**Comment by mungewell** [comment]
It's good to see so many new 'faces' here... let's party!

@PomanoB This is somewhat of my concern, that the use of a service will increase the 'investment' of trying out the PSVR and that being too deeply installed will 'fight' with the PSMoveService (and others) when they get proper support.

I believe that SteamVR will be better served once camera support is available to provide full tracking support, and that camera is probably out of scope for PSVRFramework. It would be really nice to be able to use a simple application to control the PSVR and do 'rotation' sensor fusion. There was talk of a TrackIR compatible UDP port.

But that's just my opinion... ;-)

**Comment by CalmBSK** [comment]
Just adding to the data -

With regards to getting 'Error 5' and the service failing to start I can confirm my Win10 install is the Home version. If I add user 'netservice' to the Administrator group then the PSVRService service starts successfully as a Network Service. Changing to logon as a Local Service I can also confirm my HMD connects and LEDs blink.

**Comment by PSVR-Customer** [comment]
Hi Gusmanb ! 

To install PSVRService, i had to run "Admin CMD" and enter "PSVRService.exe /i" it works like a charm

_Maybe more simple for everybody to make a batch script like Setup.cmd with Administrator Privileges Shortcut_

I have to try the new PSVRToolbox, i will edit later

**Comment by mjd1** [comment]
Hi,

Windows 7 Ultimate:

I have the same permission problem starting the service until I change it to local system account, otherwise service works okay.  The first time I started the toolbox it was 'waiting for vr', quit and restarted toolbox and it connected to the service okay.  Everything seemed to work fine.

Standalone toolbox worked okay.

Linux:
Same hardware as Windows test.  Debian/sid.
$ mono --version
Mono JIT compiler version 4.6.1 (Stable 4.6.1.5/ef43c15 Wed Oct 12 09:10:37 UTC 2016)
$ sudo mono-service --debug ./PSVRService.exe
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at PSVRFramework.PSVRDevice.Blink (System.Object state) [0x0001a] in <6f031872736f4b499434c9d7c43c027f>:0 
  at System.Threading.Timer+Scheduler.TimerCB (System.Object o) [0x00007] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00019] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00096] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 

No apparent response from the headset.  Also after the service fails to start it won't run a second time until the mono-service lock file is manually removed:
sudo rm /tmp/PSVRService.exe.lock

**Comment by Frogburt** [comment]
I git cloned the psvr_server branch and compiled in VS2015. Ran psvrservice /I with elevated permissions to install then ran PSVRToolbox.exe and it worked as expected. Win10Pro. Cant wait for SteamVR :-)

**Comment by gusmanb** [comment]
Thanks for the tests guys.

I will change the service to Local System as it seems it's more compatible. Also I will test with Linux along this week, by the exceptions and behavior it seems it receives data before the controller has been created, so just a minimal code changes will make it work as expected.

@PSVR-Customer No batch at all, the idea is the toolbox installer asks which type of usage you want (serviced or standalone) and it will take care of the set up, the current mechanism is just for these tests.

@mungewell I will try an idea I had to see if I can do the standalone toolbox a full server. But beware, the service will shutdown the device when the OS will reboot to avoid breaking Windows, the toolbox will not have that option, so is feasible you end with a broken Windows (as I said, not sure why but sometimes at boot, depending on the device's state, it hangs windows at boot and keeps it in a reboot loop, it happens in two machines I have, so not sure if it's just on these two machines or it happens on any machine, also this can be happening only on V2.0 devices as mine is updated).

**Comment by gusmanb** [comment]
Hey guys, new toolbox.

It has the full server integrated to allow remote connections when working on standalone mode. Also it *should* solve the problem at Linux, not 100% sure as for now I don't have a Linux box.

[PSVRToolbox.zip](https://github.com/gusmanb/PSVRFramework/files/617237/PSVRToolbox.zip)

**Comment by mjd1** [comment]
Hi gusmanb,

I'm running 'sudo mono ./PSVRToolbox.exe' with standalone mode enabled.  On startup the LEDs blink and the display blanks briefly but the toolbox still shows 'waiting for PS VR'.

**Comment by gusmanb** [comment]
@mjd1 Ok, at least it doesn't throws an exception, that problem has been corrected. I will try it along this week.
Thanks for the report.

**Comment by mungewell** [comment]
I can confirm the behaviour of that last zip. PSVRToolbox does flash the PSVR LEDs on startup, but does not 'find' PSVR.

Interestingly the settings do not save unless 'Mono complete' is installed, and even then the 'Run on Startup' option can not save - does not seen to be included in settings.json - or in binary looking at 'strings PSVRToolbox'.

**Comment by gusmanb** [comment]
D'oh... how bad is to be tired. I prepared last night a surprise, wrote the post, uploaded files, wrote a very funny explanation and... powered off my computer without pressing send. ROFL.

Ok guys, so, here it is, the toolbox with mouse emulation. It's still in it's early stages but it works (on Windows, on Linux besides making the toolbox work I need to program an interface to udevice to emulate the mouse and on Windows I must use mouse_move as now I'm using just Cursor.SetPosition and it's not compatible with DirectInput).

You will see six parameters on the toolbox, screen size, distance, resolution and smoothing. Width and Height represent the screen size, and distance is the distance from the viewer to the screen. This can be in any unit, centimeters, meters, inches, miles, pixels and even if you like it, midichlorians :D, while these three parameters are expressed in the same unit it should work. Resoultion is that, the screen resolution in pixels (sorry, no, no midichlorians support yet) and smoothing is an easing factor ranging from 0.0000001 to 1 being 1 no easing.

To use the mouse first of all ensure it's calibrated correctly, get the device steady and calibrate it, and then after configuring the params click the checkbox, you will see how the muse jumps to the center and then the mouse will start working. There's no support for saving settings nor options to bind it, it's the first version and needs to be finished.

Before you ask:

  *1* The toolbox does not have (*any feature you want to request*).

It's a test, so it does what it's described before, if you have some more questions, go to *1* as it should answer your questions XD. When I get this right on Windows and Linux I will add a configuration system so the parameters will be configured looking at concrete places of the screen to make it easy to use even if you don't konw the right parameters.

@mungewell Try this on Windows, it's very strange it doesn't works, maybe you have the service installed? if the service is installed and you use the toolbox on standalone mode it will not find the device. Also, the start up option is correct, it works only on Windows, it's not a config option, it's a registry entry, on WIndows it registers itself to be run on start up, on Linux I have no idea how the same can be done, any info about this (how to start a program when Gnome or KDE starts) would be apreciated.

[PSVRToolbox.zip](https://github.com/gusmanb/PSVRFramework/files/618636/PSVRToolbox.zip)
[PSVRService.zip](https://github.com/gusmanb/PSVRFramework/files/618637/PSVRService.zip)

**Comment by Frogburt** [comment]
The mouse emulation works fine for me (Win10Pro). I notice it disables the normal mouse, not sure whether it would be useful to accept movements from this also,  or to have a way to toggle between?

**Comment by MaVe-64** [comment]
Mouse emu works ok on W7 64b Home. 
Keep reading 1. I'm in a loop... ;-P

**Comment by gusmanb** [comment]
@Frogburt It doesn't disables the mouse, but as it's a "look at", unless your head is 100% stable (impossible) it will move the mouse to the computed possition, so the mouse movements seem to be blocked. Try this: leave the hmd on a table, recalibrate and recenter, ou will be able to move freely the mouse, then move the headset, it will place the mouse at it's computed place. If you leave it again stable you will be able to move it again. I can implement a second mode for the mouse emulator, instead of being absolute make it relative, but I think it's not worth it, if some real cases of use arise then I would think about it.

@MaVe-64 Questions.Clear(); break;   ;)

**Comment by PSVR-Customer** [comment]
I have tested the last PSVRToolbox & PSVRService they works as intended on winusb drivers (just need to unplug the usb)..
The Mouse emulation work as intended on windows, but dont works on games. 
If we can get a keybind for the mouse emulation too please, when you fix it for games (DirectX only ? OpenGL ok ?) 
Otherwise, if you could bind from the rigth stick of standard usb xbox joypad for PC, you should get the Headtracking in games. Pinnacle Game profiler could help for that like they do for VLC ? Dont know more sorry...
I guess every pc gamers have a usb joypad so... best choice no ? Besides we can keep the mouse emulation to navigate in interface games, it works when you enter Menu or inventory, ...

**Comment by electromat94** [comment]
I had to Install the services with the Admin command and set my ''Log on as'' to Local (in the properties of Playstation VR Server Services). Every things works fine..exept that When in cinematic mode, launching a game with Tridef 3D, and click on Enable VR, the white circle dots is displayed then black screen... Am I the only one? Or maybe the code is not updated yet for that feature?

**Comment by gusmanb** [comment]
@electromat94 Ensure your resolution is exactly 1920x1080 and if you have an AMD card set the pixel format to RGB 4:4:4.

**Comment by gusmanb** [comment]
As the mouse system seems to work OK i will try to create the set-up system and the final implementation this week, I've been extremely busy on my work so had not much spare time during the past days.

**Comment by PSVR-Customer** [comment]
@gusmanb ok, Big Thanks ! no problem... we can wait ;)
Anyway that take a big time to let us associate keybinds to HeadOrientation ? Maybe we could use Xpadder (Xpadder.com) keybinds from USB Joypad to use Mouse Axis too ... have to try it
But IRL First ! :)

**Comment by electromat94** [comment]
Thanks @gusmanb, I have a AMD card and I set it to RGB 4:4:4 . Now it's working!

**Comment by mungewell** [comment]
Hi all,
It's obviously been rather quiet around here for a while... wanted to check in and comment that I'll be travelling for work for the next week.

Do 'we' have plans for some more builds soon? I get that people are busy this time of year, but we want to keep at least some forward motion (otherwise Trinus will get all the glory ;-).
Simon

**Comment by MaVe-64** [comment]
Yeah, it would be great to have all of this excellent work  working with SteamVR. You guys can do better than Trinus :-)

**Comment by gusmanb** [comment]
Trinus is using my code and I'm on contact with the developer, so no worries at all. Sorry if I'm offline these days, I'm finishing a very big system so I will be extremely busy until new year...

**Comment by PSVR-Customer** [comment]
Well... The TrinusPSVR is not a free software like yours ... Some guys already paid to unlock the app...
Dont like this... Hope you get soon a better app to stop him ! 
Anyway he make a good job for SteamVR, still alpha... but working fine with the mouse (no focus) maybe the mouse need a deadzone like joystick ... or curve to manage mouse acceleration... dont know

Take your time, we will promote your app ! ;)

Have good christmas, see you later

**Comment by cercata** [comment]
That's cool you guys colaborating !!!! :)

I don't mind to pay for a polished product with continued support !!!!

The main isssue now is eliminate drifting ... but I don't know if it will be posible without the camera, Sony was unable to do so ...

**Comment by PSVR-Customer** [comment]
Hi Gusmanb

We wish you an Happy new year ;)

From PSVR Hack Reddit Community, we would like to know if you could give us some news about the state of PSVRToolbox (Issue about Driver conflict with TrinusPSVR) & VR Player with Gyro Mouse (Audio Bug).

Thanks for your response :)

**Comment by overdoozedj** [comment]
TO : gusmanb

For beginning ? In first ? To ALL GUYS !? : HAPPY NEW YEAR 2018 !!!!!! 

Hi Sir ^^, ALL MY CONGRADULATIONS for you're work with this app !? 

I'm OK to be a Tester ! I have The PS VR and I've downloaded : 

PSVRService.zip
and
PSVRToolbox.zip

So ? I don't like LINUX......did you have the same app on Windows ? 

Working For X86 or X64 Chipset ? 

So !? I've Signed up into this FORUM and I will get back here many times so don't worry I'm going back here ^^

PS : Sorry for my bad English ......I'm speaking French...........
